File: crates/oharness-loop/examples/hello_scripted.rs
//! `hello_scripted` — the "first agent in 10 lines" example.
//!
//! Builds a minimal [`Agent`] wired with a scripted Llm (no real API
//! calls, no cost), the shipped `fs` tool kit, and the default
//! `ReactLoop`. Runs a single turn and prints the resulting
//! termination + final assistant message.
//!
//! Run with:
//!
//! ```bash
//! cargo run --example hello_scripted -p oharness-loop
//! ```
//!
//! Intent: this is the entry-point demo users read first, so it
//! deliberately avoids any of the larger pieces (critics, budgets,
//! trajectory files, middleware, reflexion). Those have their own
//! examples in the same directory.

use async_trait::async_trait;
use oharness_core::{
    CompletionRequest, CompletionResponse, Content, LlmCapabilities, ModelId, StopReason, Task,
    Usage,
};
use oharness_llm::{ChunkStream, Llm, LlmError};
use oharness_loop::{Agent, ReactLoop};
use oharness_tools::fs::FsToolSet;
use std::sync::Arc;

/// A scripted `Llm` that returns a one-turn canned response. Real
/// adapters (`AnthropicLlm`, `OpenAiLlm`, …) slot in here without
/// any other changes to the agent.
struct ScriptedLlm;

#[async_trait]
impl Llm for ScriptedLlm {
    fn name(&self) -> &str {
        "scripted"
    }

    fn capabilities(&self) -> LlmCapabilities {
        LlmCapabilities::default()
    }

    async fn complete(&self, _req: CompletionRequest) -> Result<CompletionResponse, LlmError> {
        Ok(CompletionResponse {
            id: "hello".into(),
            model: ModelId::new("scripted-example"),
            content: vec![Content::text(
                "Hello from open-harness! Running against a scripted LLM; the \
                 trajectory of this run is captured via the default middleware \
                 stack.",
            )],
            stop_reason: StopReason::EndTurn,
            usage: Usage::default(),
        })
    }

    async fn stream(&self, _req: CompletionRequest) -> Result<ChunkStream, LlmError> {
        Err(LlmError::Unsupported("stream"))
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let agent = Agent::builder()
        .with_llm(Arc::new(ScriptedLlm))
        .with_tools(Arc::new(FsToolSet::new()))
        .with_loop(Box::new(ReactLoop::new()))
        .with_max_turns(3)
        .build()?;

    let outcome = agent.run(Task::new("say hello")).await?;

    println!("Termination: {:?}", outcome.termination);
    println!(
        "Turns: {}, tool calls: {}",
        outcome.usage.turns, outcome.usage.tool_calls
    );
    if let Some(oharness_core::Message::Assistant { content, .. }) = outcome.final_messages.last() {
        for c in content {
            if let Content::Text { text } = c {
                println!("Assistant: {text}");
            }
        }
    }
    Ok(())
}

File: crates/oharness-py/examples/hello_scripted.py
"""hello_scripted — the "first Python agent in 10 lines" example.

Builds a minimal agent wired with a scripted `Llm` (no real API
calls, no cost), the shipped `FsToolSet`, and the default
`ReactLoop`. Runs a single turn and prints the termination + final
assistant message.

Build + install the `oharness` wheel first:

    cd crates/oharness-py
    maturin develop --release

Then run:

    python examples/hello_scripted.py
"""

import json

import oharness


class HelloLlm:
    """A scripted `Llm` that returns one canned response.

    The Python class just needs a `complete(req_json) -> str`
    method. `req_json` is a JSON-encoded `CompletionRequest`
    (which this example doesn't need to inspect); the return is a
    JSON-encoded `CompletionResponse`.

    Real adapters (`AnthropicLlm`, `OpenAiLlm`, …) slot in here
    without any other changes to the agent.
    """

    def complete(self, req_json: str) -> str:
        return json.dumps({
            "id": "hello-1",
            "model": "scripted-example",
            "content": [{
                "type": "text",
                "text": (
                    "Hello from open-harness! Running against a scripted LLM; "
                    "the trajectory of this run is captured via the default "
                    "middleware stack."
                ),
            }],
            "stop_reason": {"kind": "end_turn"},
            "usage": {"tokens_input": 3, "tokens_output": 20},
        })


def main() -> None:
    llm = oharness.PyLlm(HelloLlm(), name="hello")
    agent = (
        oharness.Agent.builder()
        .with_llm(llm)
        .with_tools(oharness.FsToolSet())
        .with_loop(oharness.ReactLoop())
        .with_max_turns(3)
        .build()
    )

    outcome = json.loads(agent.run(oharness.Task("say hello")))

    print(f"Termination: {outcome['termination']}")
    print(
        f"Turns: {outcome['usage']['turns']}, "
        f"tool calls: {outcome['usage']['tool_calls']}"
    )
    for msg in outcome["final_messages"]:
        if msg.get("role") != "assistant":
            continue
        for block in msg.get("content", []):
            if block.get("type") == "text":
                print(f"Assistant: {block['text']}")


if __name__ == "__main__":
    main()
